### ISSUE TITLE
Refactor out the Query language from RocksDB

### ISSUE THREAD

**Issue by cetra3** [pipedown/noise#38]
I have a use case to match JSON responses from a variety of RESTful APIs and other sources. I'd like to be able to use the query language to parse out a single JSON document or even a JSON array.

Is there a way to do this currently?  From the source it looks like it has to be in RocksDB to query it.

**Comment by vmx** [comment]
@cetra, that is correct. We use RocksDB for persistence. So you would insert your JSON into Noise and then do a query on the data.

**Comment by vmx** [comment]
@cetra3 As there's no plan to remove the RocksDB dependency, is it OK if I close the issue?

**Comment by cetra3** [comment]
@vmx, sure. 

It would be nice to support some other backends (such as postgres + jsonb) in the future, but I understand it would be a huge undertaking.

**Comment by Damienkatz** [comment]
We need to do this. The parser should be able to generate AST representation of the query, and from that the rocksdb backend query engine would be generated, or a streaming engine, etc.

**Comment by vmx** [comment]
If someone wants to tackle the AST generation, I propose using a parser combinator. It seemed liked a clean/maintainable approach when I did some prototyping.

**Comment by cetra3** [comment]
I've used POM before for a few things.  I'd like to help but I'm not sure where to start

**Comment by vmx** [comment]
I've started a parser using [LALRPOP](https://github.com/nikomatsakis/lalrpop/) (opposed to me recommendation of doing it with a parser combinator). The code is available at https://github.com/vmx/noise-lalrpop.
